### PR TITLE
GH-37434: [C++] IO: Refactor BufferedInputStream::Read for small input

### LINKED ISSUE #37434: IO: BufferedInputStream would trigger small IO when issue small IO in edge of buffer
### Describe the enhancement requested

```c++
  Result<int64_t> Read(int64_t nbytes, void* out) {
    if (ARROW_PREDICT_FALSE(nbytes < 0)) {
      return Status::Invalid("Bytes to read must be positive. Received:", nbytes);
    }

    if (nbytes < buffer_size_) {
      // Pre-buffer for small reads
      RETURN_NOT_OK(BufferIfNeeded());
    }

    if (nbytes > bytes_buffered_) {
      // Copy buffered bytes into out, then read rest
      memcpy(out, buffer_data_ + buffer_pos_, bytes_buffered_);

      int64_t bytes_to_read = nbytes - bytes_buffered_;
      if (raw_read_bound_ >= 0) {
        bytes_to_read = std::min(bytes_to_read, raw_read_bound_ - raw_read_total_);
      }
      ARROW_ASSIGN_OR_RAISE(
          int64_t bytes_read,
          raw_->Read(bytes_to_read, reinterpret_cast<uint8_t*>(out) + bytes_buffered_));
      raw_read_total_ += bytes_read;

      // Do not make assumptions about the raw stream position
      raw_pos_ = -1;
      bytes_read += bytes_buffered_;
      RewindBuffer();
      return bytes_read;
    } else {
      memcpy(out, buffer_data_ + buffer_pos_, nbytes);
      ConsumeBuffer(nbytes);
      return nbytes;
    }
  }
```

If we Set BufferSize == 100k, and read 3k bytes per IO. When we read the 34 times, the IO would be `(99k, 102k]`

In `Read`, it will read buffered `(99k, 100k]`, issue IO for `(100k, 102k]`. Rather than `(100k, 202k]`. Is this expected?

### Component(s)

C++

### PR DESCRIPTION
**Pull request by mapleFU**
### Rationale for this change

If we Set BufferSize == 100k, and read 3k bytes per IO. When we read the 34 times, the IO would be (99k, 102k]

In Read, it will read buffered (99k, 100k], issue IO for (100k, 102k]. Rather than (100k, 200k]. 

### What changes are included in this PR?

Refactor `BufferedInputStream::Read` to optimize small IO.

### Are these changes tested?

Already has tests?

### Are there any user-facing changes?

User might get io-pattern changed. It can be optimization or downgrade.

* Closes: #37434

### CODE REVIEW

**Comment by mapleFU** [comment]
cc @pitrou @lidavidm Would you mind take a look?

File: cpp/src/arrow/io/buffered.cc [hunk 1]
```diff
@@ -373,33 +378,35 @@ class BufferedInputStream::Impl : public BufferedBase {
...
-    if (nbytes < buffer_size_) {
-      // Pre-buffer for small reads
-      RETURN_NOT_OK(BufferIfNeeded());
+    // 1. First consume pre-buffered data.
+    int64_t pre_buffer_copy_bytes = std::min(nbytes, bytes_buffered_);
+    if (pre_buffer_copy_bytes > 0) {
+      memcpy(out, buffer_data_ + buffer_pos_, pre_buffer_copy_bytes);
+      ConsumeBuffer(pre_buffer_copy_bytes);
+    }
+    int64_t remain_bytes = nbytes - pre_buffer_copy_bytes;
```
Comment (pitrou) [inline cpp/src/arrow/io/buffered.cc:387]: Nit: call this `remaining_bytes`

File: cpp/src/arrow/io/buffered.cc [hunk 2]
```diff
@@ -373,33 +378,35 @@ class BufferedInputStream::Impl : public BufferedBase {
       return Status::Invalid("Bytes to read must be positive. Received:", nbytes);
     }
 
-    if (nbytes < buffer_size_) {
-      // Pre-buffer for small reads
-      RETURN_NOT_OK(BufferIfNeeded());
+    // 1. First consume pre-buffered data.
```
Comment (pitrou) [inline cpp/src/arrow/io/buffered.cc:381]: Should you first ensure that `nbytes` does not go past `raw_read_bound_`?

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks @mapleFU ! This looks good on the principle. Some comments below.

Also, perhaps we can add tests for this? `TrackedRandomAccessFile` allows recording the exact Read calls on the raw stream, so you can easily check the different read patterns.

**Comment by mapleFU** [comment]
@pitrou I've added some tests and fix the bound

File: cpp/src/arrow/io/buffered_test.cc [hunk 3]
```diff
@@ -672,5 +671,123 @@ TEST_F(TestBufferedInputStreamBound, BufferExactlyExhausted) {
...
+// Read bytes greater than buffer_size would not buffer.
+TEST_F(TestBufferedInputStreamChunk, NotBufferLargeRead) {
+  int64_t buffer_size = 5;
+  MakeExample(buffer_size);
+
+  ASSERT_OK_AND_ASSIGN(auto buf, buffered_->Read(6));
+  EXPECT_EQ(6, buf->size());
+  EXPECT_EQ(1, tracked_->num_reads());
+  EXPECT_EQ(0, buffered_->bytes_buffered());
+  EXPECT_EQ(6, tracked_->bytes_read());
```
Comment (pitrou) [inline cpp/src/arrow/io/buffered_test.cc:718]: We could instead check the exact read ranges (untested):
```suggestion
  EXPECT_EQ(0, buffered_->bytes_buffered());
  EXPECT_EQ(tracked_->get_read_ranges(), {ReadRange{0, 6}});
```

Comment (pitrou) [reply cpp/src/arrow/io/buffered_test.cc:718]: Same suggestion below.

File: cpp/src/arrow/io/buffered.cc [hunk 4]
```diff
@@ -373,33 +378,42 @@ class BufferedInputStream::Impl : public BufferedBase {
...
+    int64_t pre_buffer_copy_bytes = std::min(nbytes, bytes_buffered_);
+    if (pre_buffer_copy_bytes > 0) {
+      memcpy(out, buffer_data_ + buffer_pos_, pre_buffer_copy_bytes);
+      ConsumeBuffer(pre_buffer_copy_bytes);
+    }
+    int64_t remaining_bytes = nbytes - pre_buffer_copy_bytes;
+    if (remaining_bytes == 0) {
+      return nbytes;
     }
```
Comment (pitrou) [inline cpp/src/arrow/io/buffered.cc:390]: This check can be removed if you rework the logic below a bit:
```c++
    DCHECK_EQ(0, bytes_buffered_);
    if (raw_read_bound_ >= 0) {
      remaining_bytes = std::min(remaining_bytes, raw_read_bound_ - raw_read_total_);
    }
    if (remaining_bytes == 0) {
      return pre_buffer_copy_bytes;
    }
```

File: cpp/src/arrow/io/buffered.cc [hunk 5]
```diff
@@ -373,33 +378,42 @@ class BufferedInputStream::Impl : public BufferedBase {
...
-          int64_t bytes_read,
-          raw_->Read(bytes_to_read, reinterpret_cast<uint8_t*>(out) + bytes_buffered_));
-      raw_read_total_ += bytes_read;
+    }
 
-      // Do not make assumptions about the raw stream position
-      raw_pos_ = -1;
-      bytes_read += bytes_buffered_;
+    // 2. Read from storage.
+    if (remaining_bytes > buffer_size_) {
```
Comment (pitrou) [inline cpp/src/arrow/io/buffered.cc:401]: Nit: no need to fill the buffer if we are going to exhaust it entirely
```suggestion
    if (remaining_bytes >= buffer_size_) {
```

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
I've added some random tests and also expanded the existing tests a bit. I'll merge if CI is green.

**Comment by pitrou** [comment]
@githutb-actions crossbow submit -g cpp

**Comment by pitrou** [comment]
@githutb-actions crossbow submit -g cpp

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

**Pull request merged**